RereleaseTimer/index.js

```diff
@@ -1,203 +1,100 @@
-    return options;
-}
+async function getRulePromise(rule, context) {
+    const owner = rule.owner;
+    const app = rule.app;
+    const sourceGroup = rule.source;
+    const destinationGroup = rule.destination;
+    const maxCrashes = rule.crashes;
+    const minInstallations = rule.installs;
+    const minSessions = rule.sessions;
+
+    // We need to make a check whether the user has specified all the properties in config.
+    if (!(owner && app && sourceGroup && destinationGroup && maxCrashes && minInstallations && minSessions)) {
+        throw new Error("ERROR: Invalid config file: missing one of the properties.");
+    }
+
+    context.log(`Processing rule for ${app} (${sourceGroup} -> ${destinationGroup})...`);
+    const releases = await appCenterApi.getRecentReleases(owner, app);
 
-function GetLatest(releases, group) {
-    for(z = 0; z < releases.length; z++) {
-        if(IsInGroup(releases[z], group)) {
-            return releases[z];
-        }
+    // Getting the latest release from the source group:
+    const release = utils.getLatestRelease(releases, sourceGroup);
+    if (!release) {
+        context.log("No releases available in source group.");
+        return;
     }
-}
 
-function IsInGroup(release, group) {
-    if(release.distribution_groups) {
-        for(i = 0; i < release.distribution_groups.length; i++) {
-            if(release.distribution_groups[i].name == group)
-                return true;
-        }
+    // If it has been already released to the destination group, stop the execution:
+    if (utils.isInGroup(release, destinationGroup)) {
+        context.log(`Latest release (${release.short_version}) has already been distributed to the destination group.`);
+        return;
     }
-    return false;
-}
 
-function FindOne(endpoint, token, owner, app) {
-    var options = BuildUrl(endpoint, token, owner, app);
-    return request(options)
-    .then(result => {
-        result = JSON.parse(result);
-        if (result) {
-            return result;
-        }
-    })
-}
+    context.log(`Checking stats for version ${release.short_version} (${release.id})...`);
+    const stats = await getStats(release, owner, app);
+    let [crashes, sessions, installations] = [...stats];
+    context.log(`Crashes Detected: ${crashes}`);
+    context.log(`Sessions (30sec-30min): ${sessions}`);
+    context.log(`Total Installations: ${installations}`);
+
+    // Proceed with the release only if 
+    // - the amount of crashes since the release has been made 
+    //   does not exceed the maximum amount specified in config file;
+    // - the amount of installations since the release has been made 
+    //   is higher than the specified in config file;
+    // - the amount of event sessions since the release has been made 
+    //   is higher than the specified in config file;
+    if (!(crashes <= maxCrashes && installations >= minInstallations && sessions >= minSessions)) {
+        context.log(`Nothing to perform.`);
```

Comment:
Don't sure if this message is right in terms of the condition above

Reply:
Got it will change